Library/Homebrew/completions.rb

```diff
@@ -0,0 +1,230 @@
+# typed: true
+# frozen_string_literal: true
+
+# Helper functions for generating shell completions.
+#
+# @api private
+module Completions
+  extend T::Sig
+
+  module_function
+
+  COMPLETIONS_DIR = (HOMEBREW_REPOSITORY/"completions").freeze
+
+  sig { void }
+  def update_shell_completions!
+    commands = Commands.commands(aliases: true).sort
+
+    (COMPLETIONS_DIR/"bash/brew").atomic_write generate_bash_completion_file(commands)
+  end
+
+  sig { params(command: String).returns(T::Boolean) }
+  def command_gets_completions?(command)
+    return false if command.start_with? "-"
+    return false if command.start_with? "cask "
+    return false if command_options(command).empty?
+
+    true
+  end
+
+  sig { params(command: String).returns(T::Array[String]) }
+  def command_options(command)
+    options = []
+    Commands.command_options(command)&.each do |option|
+      next if option.blank?
+
+      name = option.first
+      if name.start_with? "--[no-]"
+        options << name.remove("[no-]")
+        options << name.sub("[no-]", "no-")
+      else
+        options << name
+      end
+    end&.compact
+    options.sort
+  end
+
+  sig { params(command: String).returns(T.nilable(String)) }
+  def generate_bash_subcommand_completion(command)
+    return unless command_gets_completions? command
+
+    <<~COMPLETION
+      _brew_#{Commands.method_name command}() {
+        local cur="${COMP_WORDS[COMP_CWORD]}"
+        case "$cur" in
+          -*)
+            __brewcomp "
+            #{command_options(command).join "\n      "}
+            "
+            return
+            ;;
+        esac
+      }
+    COMPLETION
+  end
+
+  sig { params(commands: T::Array[String]).returns(T.nilable(String)) }
+  def generate_bash_completion_file(commands)
+    completion_functions = commands.map do |command|
+      generate_bash_subcommand_completion command
+    end.compact
+
+    subcommand_function_mappings = commands.map do |command|
+      next unless command_gets_completions? command
+
+      "#{command}) _brew_#{Commands.method_name command} ;;"
+    end.compact
+
+    <<~COMPLETION
+      # Bash completion script for brew(1)
```

Comment:
Could put some of this into an `erb` file and inject only the dynamic bits?

Reply:
Yes, this is exactly what I was looking for!